pytorch_lightning/trainer/trainer.py

```diff
@@ -2026,6 +2026,22 @@ def should_rank_save_checkpoint(self) -> bool:
     def num_nodes(self) -> int:
         return getattr(self.strategy, "num_nodes", 1)
 
+    @property
+    def device_ids(self) -> List[int]:
+        """List of device indexes."""
+        devices = getattr(self.strategy, "parallel_devices", [self.strategy.root_device])
+        device_ids = []
+        for idx, device in enumerate(devices):
+            if isinstance(device, torch.device):
+                device_ids.append(device.index or idx)
+            elif isinstance(device, int):
+                device_ids.append(device)
+        return device_ids
+
+    @property
+    def num_devices(self) -> int:
+        return len(self.strategy.parallel_devices) if isinstance(self.strategy, ParallelStrategy) else 1
```

Comment:
did the docstrings get lost? please clarify that these properties are per node
```suggestion
    def device_ids(self) -> List[int]:
        """List of device indexes used per node."""
        devices = getattr(self.strategy, "parallel_devices", [self.strategy.root_device])
        device_ids = []
        for idx, device in enumerate(devices):
            if isinstance(device, torch.device):
                device_ids.append(device.index or idx)
            elif isinstance(device, int):
                device_ids.append(device)
        return device_ids

    @property
    def num_devices(self) -> int:
        """ Number of devices the trainer uses per node. """
        return len(self.strategy.parallel_devices) if isinstance(self.strategy, ParallelStrategy) else 1
```

Reply:
also shouldn't these properties be kept in sync? why isn't `num_devices` implemented like this?
```py
    return max(len(self.device_ids), 1)
```
?

Reply:
I received a comment previously about `self.device_ids` requiring additional computations. 
https://github.com/PyTorchLightning/pytorch-lightning/pull/12151#discussion_r817486457

I also add a test case to asserting `self.num_devices == len(expected_device_ids) == len(self.device_ids)`

I'm okay with both weighing more towards logic in sync or efficient computations.